### ISSUE TITLE
Github action sporadic failure with spark

### ISSUE THREAD

**Issue by nalinigans** [GenomicsDB/GenomicsDB#151]
Some [builds](https://github.com/GenomicsDB/GenomicsDB/actions/runs/731278279) fail with `sparkDriver` not able to find a random free port. Most of the time, restarting the build works. But just wondering if you can take a look.

```
2021-04-11T18:38:04.1426794Z Sanity test with query.json : spark-submit --master local --deploy-mode client --total-executor-cores 1 --executor-memory 512M  --conf "spark.executor.extraJavaOptions=" --conf "spark.driver.extraJavaOptions=" --class org.genomicsdb.spark.api.GenomicsDBSparkBindings /home/runner/work/GenomicsDB/GenomicsDB/build/target/genomicsdb-x.y.z.test-allinone.jar /tmp/tmp69y8hzf7/sanity_test/t0_1_2.json /tmp/tmp69y8hzf7/sanity_test/query.json failed
2021-04-11T18:38:04.1429971Z Stdout:
2021-04-11T18:38:04.1431279Z Stderr: 21/04/11 18:38:03 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
2021-04-11T18:38:04.1432436Z 21/04/11 18:38:03 INFO SparkContext: Running Spark version 3.0.1
2021-04-11T18:38:04.1433443Z 21/04/11 18:38:03 INFO ResourceUtils: ==============================================================
2021-04-11T18:38:04.1437692Z 21/04/11 18:38:03 INFO ResourceUtils: Resources for spark.driver:
2021-04-11T18:38:04.1438166Z
2021-04-11T18:38:04.1438643Z 21/04/11 18:38:03 INFO ResourceUtils: ==============================================================
2021-04-11T18:38:04.1439513Z 21/04/11 18:38:03 INFO SparkContext: Submitted application: GenomicsDB API Experimental Bindings
2021-04-11T18:38:04.1440442Z 21/04/11 18:38:03 INFO SecurityManager: Changing view acls to: runner
2021-04-11T18:38:04.1441398Z 21/04/11 18:38:03 INFO SecurityManager: Changing modify acls to: runner
2021-04-11T18:38:04.1442177Z 21/04/11 18:38:03 INFO SecurityManager: Changing view acls groups to:
2021-04-11T18:38:04.1442970Z 21/04/11 18:38:03 INFO SecurityManager: Changing modify acls groups to:
2021-04-11T18:38:04.1444518Z 21/04/11 18:38:03 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(runner); groups with view permissions: Set(); users  with modify permissions: Set(runner); groups with modify permissions: Set()
2021-04-11T18:38:04.1446748Z 21/04/11 18:38:04 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
...
2021-04-11T18:38:04.1456342Z 21/04/11 18:38:04 WARN Utils: Service 'sparkDriver' could not bind on a random free port. You may check whether configuring an appropriate binding address.
2021-04-11T18:38:04.1472401Z java.net.BindException: Cannot assign requested address: Service 'sparkDriver' failed after 16 retries (on a random free port)! Consider explicitly setting the appropriate binding address for the service 'sparkDriver' (for example spark.driver.bindAddress for SparkDriver) to the correct binding address.
2021-04-11T18:38:04.1475406Z  at sun.nio.ch.Net.bind0(Native Method)
2021-04-11T18:38:04.1476149Z  at sun.nio.ch.Net.bind(Net.java:461)
2021-04-11T18:38:04.1476956Z  at sun.nio.ch.Net.bind(Net.java:453)
...
```

**Comment by mlathara** [comment]
I'll take a peek

**Comment by mlathara** [comment]
closed here: https://github.com/GenomicsDB/GenomicsDB/pull/153

**Comment by nalinigans** [comment]
Would using `https://github.com/marketplace/actions/setup-apache-spark` be an option at least for the main builds> The supported versions are only for Ubuntu and for 
- Apache Spark versions 3.0.2 and 3.1.1
- Hadoop version 3.2

**Comment by jacmarjorie** [comment]
I used this before, and was able to get it to work with spark2 by specifying the md5 for spark 2.4.x download. Though it looks like they removed that as a parameter in the action.yml. I've tried to recover this in the commits somewhere with no luck.

@nalinigans do we only see sporadic behavior with spark3? or spark2 as well?

**Comment by nalinigans** [comment]
@jacmarjorie I have seen this issue with spark2 as well.

**Comment by jacmarjorie** [comment]
ok, at a quick look the only issue I see with specifying spark2 is that this like will fail:

https://github.com/vemonet/setup-spark/blob/main/src/setup-spark.ts#L13

`https://www.apache.org/dyn/closer.lua/spark/spark-${sparkVersion}/spark-${sparkVersion}-bin-hadoop${hadoopVersion}.tgz`

if you try the link with a spark2 version, like 2.4.7, the link does not exist. Could put in a PR here though to make calling out to this URL more flexible to older versions.

**Comment by nalinigans** [comment]
A PR there sounds good, @jacmarjorie. Please do file.

**Comment by nalinigans** [comment]
Is this `vemonet` the same one as in the marketplace? Also, we may be able to look at the code and see what they are doing with respect to `spark_local_ip`.

**Comment by jacmarjorie** [comment]
I see nothing special with the spark install in that repo, just downloads it and sets env vars. At the bottom of usage in the github actions link you posted - click on actions.yaml and you will see the code base. Perhaps our hadoop install etc is messing things up. I'll test this out and see what happens.

**Comment by nalinigans** [comment]
Yes, they are only doing a tar extract and creating a link, thats all. We do some gimmicks to allow passwordless `ssh localhost` for Hadoop - all this recommended in the Hadoop docs btw, maybe this is an issue. Also, this error is sporadic and started in the last month.

**Comment by nalinigans** [comment]
Seems to have been fixed by #154. Closing for now.
